modules/renter/proto/contractset.go

```diff
@@ -22,11 +22,34 @@ type ContractSet struct {
 	mu        sync.Mutex
 }
 
-// Len returns the number of contracts in the set.
-func (cs *ContractSet) Len() int {
+// Acquire looks up the contract with the specified FileContractID and locks
+// it before returning it. If the contract is not present in the set, Acquire
+// returns false and a zero-valued RenterContract.
+func (cs *ContractSet) Acquire(id types.FileContractID) (modules.RenterContract, bool) {
 	cs.mu.Lock()
-	defer cs.mu.Unlock()
-	return len(cs.contracts)
+	safeContract, ok := cs.contracts[id]
+	cs.mu.Unlock()
+	if !ok {
+		return modules.RenterContract{}, false
+	}
+	safeContract.mu.Lock()
+	return safeContract.RenterContract, true
+}
+
+// Delete removes a contract from the set. The contract must have been
+// previously acquired by Acquire. If the contract is not present in the set,
+// Delete is a no-op.
+func (cs *ContractSet) Delete(contract modules.RenterContract) {
+	cs.mu.Lock()
+	safeContract, ok := cs.contracts[contract.ID]
+	if !ok {
+		cs.mu.Unlock()
```

Comment:
your cs.mu.Unlock()s can occur in a defer here.

Reply:
`cs.mu` needs to be unlocked before `safeContract.mu`